pkg/manager/volumes/vol_compare_utils.go

```diff
@@ -0,0 +1,385 @@
+
+	case v1alpha1.TiKVMemberType:
+		defaultScName = tc.Spec.TiKV.StorageClassName
+		d := DesiredVolume{
+			Name:             v1alpha1.GetStorageVolumeName("", mt),
+			Size:             getStorageSize(tc.Spec.TiKV.Requests),
+			StorageClassName: defaultScName,
+		}
+		desiredVolumes = append(desiredVolumes, d)
+
+		storageVolumes = tc.Spec.TiKV.StorageVolumes
+
+	case v1alpha1.TiFlashMemberType:
+		for i, claim := range tc.Spec.TiFlash.StorageClaims {
+			d := DesiredVolume{
+				Name:             v1alpha1.GetStorageVolumeNameForTiFlash(i),
+				Size:             getStorageSize(claim.Resources.Requests),
+				StorageClassName: claim.StorageClassName,
+			}
+			desiredVolumes = append(desiredVolumes, d)
+		}
+
+	case v1alpha1.TiCDCMemberType:
+		defaultScName = tc.Spec.TiCDC.StorageClassName
+		storageVolumes = tc.Spec.TiCDC.StorageVolumes
+
+	case v1alpha1.PumpMemberType:
+		defaultScName = tc.Spec.Pump.StorageClassName
+		d := DesiredVolume{
+			Name:             v1alpha1.GetStorageVolumeName("", mt),
+			Size:             getStorageSize(tc.Spec.Pump.Requests),
+			StorageClassName: defaultScName,
+		}
+		desiredVolumes = append(desiredVolumes, d)
+	default:
+		return nil, fmt.Errorf("unsupported member type %s", mt)
+	}
+
+	for _, sv := range storageVolumes {
+		if quantity, err := resource.ParseQuantity(sv.StorageSize); err == nil {
+			d := DesiredVolume{
+				Name:             v1alpha1.GetStorageVolumeName(sv.Name, mt),
+				Size:             quantity,
+				StorageClassName: sv.StorageClassName,
+			}
+			if d.StorageClassName == nil {
+				d.StorageClassName = defaultScName
+			}
+
+			desiredVolumes = append(desiredVolumes, d)
+
+		} else {
+			klog.Warningf("StorageVolume %q in %s/%s .spec.%s is invalid", sv.Name, tc.GetNamespace(), tc.GetName(), mt)
+		}
+	}
+
+	if scLister != nil {
+		for i := range desiredVolumes {
+			if desiredVolumes[i].StorageClassName != nil {
+				sc, err := getStorageClass(desiredVolumes[i].StorageClassName, scLister)
+				if err != nil {
+					return nil, fmt.Errorf("cannot get sc %s", *desiredVolumes[i].StorageClassName)
+				}
+				desiredVolumes[i].StorageClass = sc
+			}
+		}
+	}
+
+	return desiredVolumes, nil
+}
+
+func isStorageClassMatched(sc *storagev1.StorageClass, scName string) bool {
+	if sc == nil {
+		// cannot get sc or sc is unset
+		return true
+	}
+	if sc.Name == scName {
+		return true
+	}
+
+	return false
+}
+
+func getStorageClass(name *string, scLister storagelister.StorageClassLister) (*storagev1.StorageClass, error) {
+	if name == nil {
+		return nil, nil
+	}
+	return scLister.Get(*name)
+}
+
+func getDesiredVolumeByName(vs []DesiredVolume, name v1alpha1.StorageVolumeName) *DesiredVolume {
+	for i := range vs {
+		v := &vs[i]
+		if v.Name == name {
+			return v
+		}
+	}
+
+	return nil
+}
+
+func getStorageSize(r corev1.ResourceList) resource.Quantity {
+	return r[corev1.ResourceStorage]
+}
+
+func ignoreNil(s *string) string {
+	if s == nil {
+		return ""
+	}
+	return *s
+}
```

Comment:
Can we use `pointer.StringDeref` instead? See `"k8s.io/utils/pointer"`

Reply:
This was old code's function so I left it untouched, but let me know if you prefer to update it with existing uses also to StringDeref

https://github.com/pingcap/tidb-operator/blob/899409c2f2858b6aa8e2f9a46db4607e32ee6fc4/pkg/manager/volumes/pvc_modifier.go#L110